Documentation/error-handling.md

```diff
@@ -0,0 +1,175 @@
+## Dialing In gRPC
+
+`grpc.Dial` is a function in the gRPC library that creates a virtual connection
+from the gRPC client to the gRPC server.  It takes a server URI ( which can
+represent the name of a logical backend service and could resolve to multiple
+actual addresses) and a list of options, and returns a `ClientConn` object that
+represents the connection to the server. The `ClientConn` contains one or more
+actual connections to real server backends and attempts to keep these
+connections healthy by automatically reconnecting to them (with an exponential
+backoff) when they break.
+
+The `grpc.Dial` function can also be configured with various options to
+customize the behavior of the client connection. For example, developers could
+use options such a WithTransportCredentials() to configure the transport
+credentials to use.
+
+While `grpc.Dial` is commonly referred to as a "dialing" function, it doesn't
+actually perform the low-level network dialing operation like a typical TCP
+dialing function would. Instead, it creates a virtual connection from the gRPC
+client to the gRPC server.
+
+`grpc.Dial` does initiate the process of connecting to the server, but it uses
+the ClientConn object to manage and maintain that connection over time. This is
+why errors encountered during the initial connection are no different from those
+that occur later on, and why it's important to handle errors from RPCs rather
+than relying on options like `FailOnNonTempDialError`, `WithBlock`, and
+`WithReturnConnectionError`.
+
+### Using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
+
+The gRPC API provides several options that can be used to configure the behavior
+of dialing and connecting to a gRPC server. Some of these options, such as
+`FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`, rely on
+failures at dial time. However, we strongly discourage developers from using
+these options, as they can introduce race conditions and result in unreliable
+and difficult-to-debug code.
+
+###  Why we discourage using `FailOnNonTempDialError`, `WithBlock`, and
+`WithReturnConnectionError`
```

Comment:
This till needs to a single line so that markdown sees the whole thing as a heading.